zerostash/src/commands/mount.rs

```diff
@@ -0,0 +1,279 @@
+//! `mount` subcommand
+
+use std::ffi::OsStr;
+use std::path::PathBuf;
+use std::sync::{mpsc, Arc, Mutex};
+use std::time::{Duration, SystemTime};
+use std::{collections::HashMap, path::Path};
+
+use std::io::Result;
+
+use crate::prelude::*;
+use fuse_mt::*;
+use infinitree::object::Reader;
+use infinitree::Infinitree;
+use nix::libc;
+use zerostash_files::{restore, Entry, Files};
+
+#[derive(Command, Debug)]
+pub struct Mount {
+    #[clap(flatten)]
+    stash: StashArgs,
+
+    #[clap(flatten)]
+    options: restore::Options,
+
+    /// The location the filesytem will be mounted on
+    #[clap(short = 'T', long = "target")]
+    mount_point: String,
+}
+
+#[cfg(unix)]
+#[async_trait]
+impl AsyncRunnable for Mount {
+    /// Start the application.
+    async fn run(&self) {
+        let stash = self.stash.open();
+
+        if let Err(e) = mount(stash, &self.options, &self.mount_point) {
+            panic!("Error = {}", e)
+        }
+    }
+}
+
+fn mount(
+    stash: Infinitree<Files>,
+    options: &restore::Options,
+    mountpoint: &str,
+) -> anyhow::Result<()> {
+    let (tx, finished) = mpsc::sync_channel(2);
+    let destroy_tx = tx.clone();
+    ctrlc::set_handler(move || tx.send(()).expect("Could not send signal on channel."))
+        .expect("Error setting Ctrl-C handler");
+
+    let filesystem = SimpleFs::open(stash, options, destroy_tx).unwrap();
+    let fuse_args = [OsStr::new("-o"), OsStr::new("fsname=zerostash")];
+
+    let fs = fuse_mt::FuseMT::new(filesystem, 1);
+
+    // Mount the filesystem.
+    let handle = spawn_mount(fs, mountpoint, &fuse_args[..])?;
+
+    // Wait until we are done.
+    finished.recv().expect("Could not receive from channel.");
+
+    // Ensure the filesystem is unmounted.
+    handle.join();
+
+    Ok(())
+}
+
+type DirPath = PathBuf;
+type FilePath = PathBuf;
+type FileContent = Vec<u8>;
+type Metadata = Arc<Entry>;
+
+struct SimpleFs {
+    destroy_tx: mpsc::SyncSender<()>,
+    stash: Infinitree<Files>,
+    file_map: HashMap<FilePath, Metadata>,
+    file_content: Mutex<HashMap<FilePath, FileContent>>,
```

Comment:
Instead of `Mutex<HashMap<K, V>>`, `infinitree` uses the `scc` crate, which implements parallel, lockless hashmaps. These mean much better speed during parallel access.